### PR TITLE
Ssl cert refreshing

### PR DESCRIPTION
**Pull request by NathanSpeidel**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -40,6 +40,11 @@
     </properties>
 
     <dependencies>
+        <dependency>
```
Comment (akshaisarma) [inline pom.xml:43]: Move to above bullet-core. This is the tests area of the pom

File: src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+import java.nio.file.Paths;
+import java.security.KeyStore;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@Slf4j
+public class CertRefreshingSslEngineFactory implements SslEngineFactory {
+    public static final String SSL_CERT_LOCATION_CONFIG = "ssl.cert.refreshing.cert.location";
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java:31]: Let's move these to KafkaConfig and add the required validations for them if present. Add default values to KafkaConfig with `: ` as values (null)

File: src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        SSLParameters sslParams = sslEngine.getSSLParameters();
+        sslParams.setEndpointIdentificationAlgorithm(endpointIdentification);
+        sslEngine.setSSLParameters(sslParams);
+
+        return sslEngine;
+    }
+
+    @Override
+    public SSLEngine createServerSslEngine(String peerHost, int peerPort) {
+        throw new RuntimeException(this.getClass().getCanonicalName() + " only supports client-side SSLEngines, it " +
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java:92]: `throw new UnsupportedOperationException("This only supports client-side SSLEngines")` The stack trace should include all the class and method names and the exception signifies why clearly

File: src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+        }
+    }
+
+    // Functions that proxy to other classes can be extended for testing
+
+    protected KeyStore getKeyStore(String jksFilePath, char[] password) throws Exception {
+        return Utils.getKeyStore(jksFilePath, password);
+    }
+
+    protected KeyStore createKeyStore(String athenzPublicCert, String athenzPrivateKey) throws Exception {
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java:162]: Replace all occurences of athenzPublicCert and athenzPrivateKey with x509PublicCert and x509PrivateKey here and elsewhere

File: src/test/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.yahoo.bullet.kafka.CertRefreshingSslEngineFactory.SSL_CERT_LOCATION_CONFIG;
+import static com.yahoo.bullet.kafka.CertRefreshingSslEngineFactory.SSL_KEY_LOCATION_CONFIG;
+import static com.yahoo.bullet.kafka.CertRefreshingSslEngineFactory.SSL_KEY_REFRESH_INTERVAL_CONFIG;
+import static org.apache.kafka.common.config.SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG;
+import static org.apache.kafka.common.config.SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG;
+import static org.mockito.Mockito.*;
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactoryTest.java]: No star imports

File: src/test/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        SSLEngine mock = factory.createClientSslEngine("peerHost", 88, "myAwesomeEndpointIdentification");
+
+        verify(mock).setEnabledCipherSuites((String[]) cipherSuites.toArray());
+        verify(mock, times(0)).setEnabledProtocols(any());
+        verify(mock).setUseClientMode(true);
+        Assert.assertEquals(mock.getSSLParameters().getEndpointIdentificationAlgorithm(), "myAwesomeEndpointIdentification");
+    }
+
+    @Test
+    public void testCreateServerSslEngine() {
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactoryTest.java:114]: Replace with expectedExceptions and expectedExceptionsRegexp

File: src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        return Utils.createKeyStore(athenzPublicCert, athenzPrivateKey);
+    }
+
+    protected KeyRefresher generateKeyRefresher(String trustStorePath, String trustStorePassword, String athenzPublicCert, String athenzPrivateKey) throws Exception {
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java:166]: Should add javadocs for all these protected methods to indicate what the override is for

File: src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    @Override
+    public Set<String> reconfigurableConfigs() {
+        return Collections.emptySet();
+    }
+
+    @Override
+    public KeyStore keystore() {
+        // As of kafka-client 2.6 this method is only used on the server-side, so we don't need to return a real
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java:114]: Let's keep it simple and throw an UnsupportedOperationException here. We can handle it if we ever need to in later versions when we go there. Should decrease branches and increase coverage

File: src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+            return createKeyStore(publicCertLocation, privateKeyLocation);
+        } catch (Exception e) {
+            log.error("Error creating keystore - this function should not be getting called on the client side.", e);
+            return null;
+        }
+    }
+
+    @Override
+    public KeyStore truststore() {
+        // As of kafka-client 2.6 this method is only used on the server-side, so we don't need to return a real
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java:126]: Ditto keystore()

File: src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java (hunk 8 above)
Comment (NathanSpeidel) [reply src/main/java/com/yahoo/bullet/kafka/CertRefreshingSslEngineFactory.java:114]: We can make a Keystore pretty easily, so I'd prefer to properly implement the functions that are expected for this class.   I tried cleaning up the comments and error message - let me know what you think.

File: src/main/java/com/yahoo/bullet/kafka/CertRefreshingSSLEngineFactory.java [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        SSLParameters sslParams = sslEngine.getSSLParameters();
+        sslParams.setEndpointIdentificationAlgorithm(endpointIdentification);
+        sslEngine.setSSLParameters(sslParams);
+
+        return sslEngine;
+    }
+
+    @Override
+    public SSLEngine createServerSslEngine(String peerHost, int peerPort) {
+        throw new RuntimeException("This only supports client-side SSLEngines, it does not support createServerSslEngine().");
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/kafka/CertRefreshingSSLEngineFactory.java:95]: UnsupportedOperationException

File: src/main/java/com/yahoo/bullet/kafka/CertRefreshingSSLEngineFactory.java [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+            return sslContext;
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to create SSLContext.", e);
+        }
+    }
+
+    /**
+     * Generate a keystore from the given jks file and password.
+     *
+     * @param jksFilePath the path to the jks file.
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/kafka/CertRefreshingSSLEngineFactory.java:162]: Capitalize the first letter of the sentence for the javadoc params and the return here and below

File: src/main/java/com/yahoo/bullet/kafka/CertRefreshingSSLEngineFactory.java [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+    protected KeyStore getKeyStore(String jksFilePath, char[] password) throws Exception {
+        return Utils.getKeyStore(jksFilePath, password);
+    }
+
+    /**
+     * Generate a KeyStore from the given public cert and private key.
+     *
+     * @param publicCertLocation the location of the public cert.
+     * @param privateKeyLocation the location of the private key.
+     * @return the KeyStore.
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/kafka/CertRefreshingSSLEngineFactory.java:176]: lowercase KeyStore -> keystore here and elsewhere

File: src/main/java/com/yahoo/bullet/kafka/CertRefreshingSSLEngineFactory.java [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+import java.nio.file.Paths;
+import java.security.KeyStore;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * This implementation of Kafka's {@link SslEngineFactory} will periodically check the certs used for SSL
+ * authentication and rebuild the KeyStore, and/or reload the Truststore if changes have been made to the
```
Comment (akshaisarma) [inline src/main/java/com/yahoo/bullet/kafka/CertRefreshingSSLEngineFactory.java:31]: lowercase Truststore -> truststore in comments

File: src/main/resources/bullet_kafka_defaults.yaml [hunk 14]
```diff
@@ -39,7 +39,28 @@ bullet.pubsub.kafka.subscriber.max.uncommitted.messages: 50
...
 bullet.pubsub.kafka.response.topic.name: "bullet.responses"
 
-# The following settings are optional. If not provided, the topics configured for requests and responses are fully used.
+# Optional settings:
+
+# The following settings can be used when "SSL" is being used to authenticate with Kafka:
+# bullet.pubsub.kafka.producer.security.protocol: "SSL"
+
+# This class will automatically refresh certs in the Kafka client producer at a configurable interval
+# bullet.pubsub.kafka.producer.ssl.engine.factory.class: "com.yahoo.bullet.kafka.CertRefreshingSslEngineFactory"
```
Comment (akshaisarma) [inline src/main/resources/bullet_kafka_defaults.yaml:48]: CertRefreshingSslEngineFactory -> CertRefreshingSSLEngineFactory

File: src/main/resources/bullet_kafka_defaults.yaml [hunk 15]
```diff
@@ -39,7 +39,28 @@ bullet.pubsub.kafka.subscriber.max.uncommitted.messages: 50
...
+# bullet.pubsub.kafka.producer.ssl.truststore.location: "my-truststore.jks"
+# bullet.pubsub.kafka.producer.ssl.truststore.password: "changeit"
+# bullet.pubsub.kafka.producer.ssl.cert.refreshing.cert.location: "my-cert.pem"
+# bullet.pubsub.kafka.producer.ssl.cert.refreshing.key.location: "my-key.pem"
+# bullet.pubsub.kafka.producer.ssl.cert.refreshing.refresh.interval.ms: "3600000"
+
+# This class will automatically refresh certs in the Kafka client consumer at a configurable interval
+# bullet.pubsub.kafka.consumer.ssl.engine.factory.class: "com.yahoo.bullet.kafka.CertRefreshingSslEngineFactory"
```
Comment (akshaisarma) [inline src/main/resources/bullet_kafka_defaults.yaml:56]: Ditto

File: src/test/java/com/yahoo/bullet/kafka/CertRefreshingSSLEngineFactoryTest.java [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+        conf.put(SSL_TRUSTSTORE_LOCATION_CONFIG, "pom.xml");
+        instantiatedFactory.configure(conf);
+        Assert.assertNotNull(instantiatedFactory.publicCertLocation);
+        Assert.assertNotNull(instantiatedFactory.privateKeyLocation);
+        Assert.assertNotNull(instantiatedFactory.truststoreLocation);
+        Assert.assertNotNull(instantiatedFactory.truststorePassword);
+    }
+
+    @Test
+    public void testCreateClientSslEngine() {
```
Comment (akshaisarma) [inline src/test/java/com/yahoo/bullet/kafka/CertRefreshingSSLEngineFactoryTest.java:129]: Ssl -> SSL here and elsewhere in this file

**Review by akshaisarma — APPROVED** [review]

**Pull request merged**